Non-counting questions

jooooosef opened the pull request:
closes #2391
fixes #2539 


Note: Since I changed the `RatingResult` (added `counts_for_grade`) you may have to `refresh_results_cache` for sites like "Evaluate" to work.

richardebeling commented on evap/staff/templates/staff_questionnaire_form.html:
```diff
@@ -58,9 +58,19 @@ <h5 class="card-title">{% translate 'Questions' %}</h5>
                                             <input class="form-check-input allow-textanswer-checkbox" id="{{ form_element.allows_additional_textanswers.id_for_label }}" name="{{ form_element.allows_additional_textanswers.html_name }}"
                                                 type="checkbox" data-keep="true" {% if form_element.allows_additional_textanswers.value %} checked{% endif %}
-                                                {% if form_element.allows_additional_textanswers.field.disabled %} disabled{% endif %} />
+                                                {% for attr_name, attr_value in form_element.allows_additional_textanswers.field.widget.attrs.items %}
+                                                    {{ attr_name }}="{{ attr_value }}"
+                                                {% endfor %} />
```
My intention here was to `include` the template from django where this is implemented. If we can't do that, for some reason, let's stick to the explicitly rendered `disabled` attribute (like you had before).

richardebeling commented on evap/evaluation/tests/test_models.py:
```diff
@@ -1139,3 +1140,37 @@ class QuestionnaireTests(TestCase):
+            )
+
+            question_heading.save(update_fields=["text_de"])
+            self.assertEqual(mock_save.call_count, 2)
```
There is [`reset_mock`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.reset_mock), which would probably make sense here (to assert two times that there was exactly the one call you expected)

richardebeling reviewed (commented):
Looking pretty good, thanks :)

niklasmohrin commented on evap/staff/templates/staff_questionnaire_form.html:
The checkboxes are not disabled for questions still in use. Updating them and revisiting the page still shows the old values (see my comment "should we swallow these errors here")

niklasmohrin commented on evap/student/templates/student_vote_questionnaire_group.html:
```diff
@@ -35,6 +35,9 @@ <h4 class="{% if not forloop.first %} mt-3 {% elif contributor %} me-auto {% end
                         {% if field|is_choice_field %}
                             <div class="col-question col-lg-4 col-xl-5 d-flex justify-content-between{% if field.errors %} choice-error{% endif %}">
                                 <span class="me-1">
+                                    {% if not field.field.widget.attrs.counts_for_grade %}
```
as far as I can tell this never occurs for the active evaluations in the test data, can you change the test data so that this shows up in for example Beautiful Data?

niklasmohrin commented on evap/results/tools.py:
line 412:
```python
    # will contain a list of question results for each contributor and one for the evaluation (where contributor is None)
    grouped_results = defaultdict(list)
    for contribution_result in get_results(evaluation).contribution_results:
        for questionnaire_result in contribution_result.questionnaire_results:
            if not questionnaire_result.questionnaire.is_dropout:  # dropout questionnaires are not counted
                grouped_results[contribution_result.contributor].extend(questionnaire_result.question_results)
```
I guess we should remove the handling for dropout here and adjust the constraints so that questions in dropout questionnaires always have `counts_for_grade=False`. @richardebeling @fekoch Does that sound about right?

If so, we then should check if it looks silly that all dropout questions have the "doesn't count for grade" marker and maybe add just a big one or so (ask Janno when the time comes)

richardebeling replied:
Not having a database constraint is not optimal, but we can still enforce this in form validation logic (clean method can enforce this), and with an assert in result computation (here). Then, we should be fine.

niklasmohrin commented on evap/results/tools.py:
line 432 calls the functions you modified, but then calculates a max over all results of rating questions; is that correct? either way, I think we need a comment that explains this, or make it clear from the code

niklasmohrin reviewed (changes requested):
okay some more concerns; this is not an easy issue :^)

jooooosef commented on evap/student/forms.py:
```diff
@@ -29,12 +29,14 @@ def from_question(cls, question: Question):
 
 class RatingAnswerField(forms.TypedChoiceField):
-    def __init__(self, widget_choices, *args, allows_textanswer=False, **kwargs):
+    def __init__(self, widget_choices, *args, allows_textanswer=False, counts_for_grade=True, **kwargs):
```
Yes since this class is only used for the from_question classmethod and the values are selected from the quesetion this is fine.

jooooosef commented on evap/evaluation/models.py:
```diff
@@ -1247,14 +1248,21 @@ class Meta:
+            self.counts_for_grade = False
             if "update_fields" in kwargs:
-                kwargs["update_fields"] = {"allows_additional_textanswers"}.union(kwargs["update_fields"])
+                kwargs["update_fields"] = {"allows_additional_textanswers", "counts_for_grade"}.union(
+                    kwargs["update_fields"]
+                )
```
`kwargs["update_fields"] |= {"allows_additional_textanswers", "counts_for_grade"}`
TypeError: unsupported operand type(s) for |=: 'list' and 'set'

jooooosef commented on evap/staff/forms.py:
```diff
@@ -935,12 +943,14 @@ class Meta:
         super().__init__(*args, **kwargs)
         if self.instance.pk and self.instance.type in [QuestionType.TEXT, QuestionType.HEADING]:
-            self.fields["allows_additional_textanswers"].disabled = True
+            self.fields["allows_additional_textanswers"].widget.attrs["disabled"] = "disabled"
```
No.
When initially loading the page, we also want to disable specific checkboxes. Instead of writing more JS this does that very easily.

richardebeling commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,147 @@
+        this.questionTable = selectOrError<HTMLTableElement>(`#${questionTableId}`);
+        this.questionnaireTypeSelect = selectOrError<HTMLSelectElement>('select[name="type"]');
+    }
+    private disableAndUncheckCheckbox = (checkbox: HTMLInputElement): void => {
```
Is there any advantage of using this syntax over the traditional syntax
```
private disableAndUncheckCheckbox(checkbox: HTMLInputElement) {}
```
? My understanding is that typescript derives a "void" return type if the function body does not contain a `return` statement.

If there's no advantage, my vote would be to use the classical syntax (@niklasmohrin am I missing something?)

jooooosef replied:
I am (green) arrow-man.
Since I used this consistently, this should be fine.

richardebeling commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,147 @@
+
+    constructor(questionTableId: string) {
+        this.questionTable = selectOrError<HTMLTableElement>(`#${questionTableId}`);
+        this.questionnaireTypeSelect = selectOrError<HTMLSelectElement>('select[name="type"]');
```
This can match any `select name="type"` element on the page, which is brittle. I suspect the select should be in the table? If yes, pass in the table as the second parameter to `selectOrError`, otherwise, we must find some other mechanism to uniquely identify this select.

jooooosef replied:
No this is the select from the questionnaire type at the very top (top questionnaire, dropout questionnaire, ...).

It's still brittle, so I changed it to its `id` which should not change as long as this form is not touched. 
Since I will write tests which will rely on the `questionnaireTypeSelect` existing, this should be fine?

richardebeling commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,147 @@
+        }
+    };
+
+    public initialize = (): void => {
```
Seems to me that initialization is the purpose of the constructor, having a method `initialize` is smelly. The single user of this class now does
```js
            const staffQuestionnaireForm = new StaffQuestionnaireForm("question_table");
            staffQuestionnaireForm.initialize();
```
Is there an advantage to this, or should this just be private and called from the constructor, or possibly even be inlined into the constructor?

richardebeling commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,147 @@
+        // Check if this is a dropout questionnaire before enabling checkboxes
+        const questionnaireType = saneParseInt(this.questionnaireTypeSelect.value);
+
+        if (questionnaireType === QUESTIONNAIRE_TYPE_DROPOUT) {
```
```suggestion

        const questionnaireType = saneParseInt(this.questionnaireTypeSelect.value);
        if (questionnaireType === QUESTIONNAIRE_TYPE_DROPOUT) {
```

richardebeling commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,133 @@
+        checkbox.disabled = false;
+    };
+
+    private disableAllCheckboxes = (checkboxes: NodeListOf<Element>) => {
```
My point was that I'd name them something like `disableAndUncheck(checkbox: HTMLInputElement)` and `disabledAndUncheckAll(checkboxes: NodeListOf<Element>)`. What do you think about that?

jooooosef commented on evap/staff/tests/test_live.py:
```diff
@@ -174,3 +175,102 @@ def test_collapse_without_editor_approved(self) -> None:
+        type_select = row.find_element(By.CSS_SELECTOR, "select[id$='-type']")
+
+        self.select_tom_select_option(type_select, str(QuestionType.TEXT))
+        self.assertTrue(
```
I added a helper for that (`assert_question_type_controls`).
Idk if thats what you expected. Also added the `*` to force the `allows_additional_textanswers_disabled` and `counts_for_grade_disabled` parameters. Is that a common practice for us here?

richardebeling commented on evap/results/tests/test_tools.py:
```diff
@@ -417,24 +419,111 @@ def test_calculate_average_course_distribution(self):
+            non_grade_distribution, (0, 0, 1, 0, 0)
+        )  # Only the counting likert question should be included
+
+        RatingAnswerCounter.objects.all().delete()
```
If you invert the order of these two sub-cases, the second sub-case doesn't have two clean up and re-create the RACs that the first one created:
* Create RACs for two non-counting questions
* assert that distributions are None
* create RACs for two counting questions
* assert that distributions are as expected

richardebeling commented on evap/results/tests/test_tools.py:
```diff
@@ -417,24 +419,111 @@ def test_calculate_average_course_distribution(self):
+
+        average_grade = distribution_to_grade(calculate_average_distribution(self.evaluation))
+
+        self.assertAlmostEqual(average_grade, 1.5)
```
Since this is exactly the contents of the test below (everything up to the distribution we compute), and the just adds the `distribution_to_grade` computation, which is completely unrelated to `counts_for_grade`, I would merge this with the test below, i.e.: Remove this test, and add two `self.assertAlmostEqual(distribution_to_grade(distribution), 1.5)` statements to the test below.

richardebeling commented on evap/results/tests/test_tools.py:
```diff
@@ -417,24 +419,111 @@ def test_calculate_average_course_distribution(self):
+        self.assertIsNone(average_grade_questions_distribution(question_results))
+        self.assertIsNone(average_non_grade_rating_questions_distribution(question_results))
+
+    def test_dropout_questionnaire_excluded_from_distribution(self):
```
nice!

richardebeling commented on evap/staff/tests/test_forms.py:
```diff
@@ -1149,3 +1150,54 @@ def test_save_makes_a_copy(self):
+        even after changing the question type from TEXT/HEADING to a Likert type.
+        Regression test for #2539.
+        """
+        questionnaire = baker.make(Questionnaire, type=Questionnaire.Type.TOP)
```
`questionnaire` is unused afterwards, just use `questionnaire__type=Questionnaire.Type.TOP` when creating the `question`.

richardebeling commented on evap/staff/tests/test_live.py:
```diff
@@ -178,6 +179,128 @@ def test_collapse_without_editor_approved(self) -> None:
+        row = self.wait.until(visibility_of_element_located((By.CSS_SELECTOR, "#question_table tbody tr")))
+        type_select = row.find_element(By.CSS_SELECTOR, "select[id$='-type']")
+
+        self.assert_question_type_controls(
```
I would pull the definition of this function inside this function body here. Then, you don't have to pass "row" and "type_select" every time, it can just use them.

I would then wiggle the names a bit to get to a point where one of these assert statements is one line, maybe something like
```python3
self.assert_question_type_allows(QuestionType.Text, additional_textanswers=False, counts_for_grade=False)
```

(note: I intentionally inverted the logic for the two attributes, because reasoning about a _disabled_ state requires one extra mental step -- make it easy for the readers)

jooooosef replied:
We could do:
```python
assert_type_allows(
            new_row, new_type_select, QuestionType.TEXT, additional_textanswers=False, counts_for_grade=False
        )
```

But in most cases it fits in one row:
```python
assert_type_allows(row, type_select, QuestionType.TEXT, additional_textanswers=False, counts_for_grade=False)
```

richardebeling commented on evap/staff/tests/test_live.py:
```diff
@@ -178,6 +179,128 @@ def test_collapse_without_editor_approved(self) -> None:
+        )
+
+    def select_tom_select_option(self, select_element, value):
+        self.selenium.execute_script(f"arguments[0].tomselect.setValue('{value}');", select_element)
```
Other tests use this as well, this should be a helper in `LiveServerTest`. My vote for the name would be `set_tomselect(instance, value)` or `set_tomselect_value(instance, value)`

richardebeling replied:
I would just do it here (as you now did), less work, simpler life

richardebeling commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,125 @@
+
+            checkboxes.forEach(checkbox => {
+                const checkboxElement = checkbox as HTMLInputElement;
+                const isCountsForGrade = checkboxElement.classList.contains("counts-for-grade-checkbox");
```
Ok, let's keep it

richardebeling reviewed (commented):
Two remaining open comments, then lgtm:
* https://github.com/e-valuation/EvaP/pull/2450#discussion_r2620988700
* https://github.com/e-valuation/EvaP/pull/2450#discussion_r2777949922

niklasmohrin reviewed (commented):
CI fails

janno42 reviewed (changes requested):
When opening the edit page of a questionnaire that cannot be edited, all fields except the checkboxes are disabled. The checkboxes should be disabled as well.

jooooosef commented on evap/static/ts/src/staff-questionnaire-form.ts:
Ah yes the logik already is in a `if editable` block, but I now added the disabling to the HTML Template

Kakadus commented:
> @jooooosef We now merged #2513 which gives some conflicts, but I don't think it will be too bad. After that, I think we are mostly ready to merge this one here too

If there are questions, I'm here to help :)

niklasmohrin commented:
(please re-request once conflicts are resolved)

jooooosef committed:
Add non-counting questions to grade calculation

jooooosef committed:
Expose questionnaire checkbox controls

jooooosef committed:
fix model and migration for counts_for_grade

jooooosef committed:
fix the grade calculation logic to use counts_for_grade

jooooosef committed:
fix frontend and logic to use counts_for_grade from question assignment

jooooosef committed:
fix checkbox disabled bug and add tests

jooooosef committed:
use assignment instead of question for student frontend

jooooosef committed:
change question to question_assignment when needed in tests

jooooosef committed:
already disbale the counts_for_grade checkbox before rendering

matches the behaviour from the allows_additional_text_answer in QuestionForm

jooooosef committed:
fix tests

jooooosef committed:
fix question assignment creation in frontendtests

jooooosef committed:
update test_data

jooooosef committed:
update minimal test_data

niklasmohrin commented on evap/staff/forms.py:
```diff
@@ -954,11 +954,25 @@ def __init__(self, *args, instance=None, **kwargs):
+            self.fields["counts_for_grade"].disabled = True  # disable only for frontend; validate in clean
 
-    def clean(self) -> None:
+    def clean(self):
```
please replace this with a suitable type instead

niklasmohrin commented on evap/staff/forms.py:
```diff
@@ -954,11 +954,25 @@ def __init__(self, *args, instance=None, **kwargs):
         super().clean()
         if not self.question_form.is_valid():
             raise forms.ValidationError([])  # invalidate this form without message; errors are shown separately
+        questionnaire = self.cleaned_data.get("questionnaire") or getattr(self.instance, "questionnaire", None)
```
why do we need both? can we just use one, or add a comment?

jooooosef replied:
No, we dont need the second one.
I was a little confused here :)

niklasmohrin commented on evap/results/tools.py:
```diff
@@ -328,7 +342,7 @@ def average_grade_questions_distribution(results):
             (unipolarized_distribution(result), result.count_sum)
             for result in results
-            if result.question.is_grade_question
+            if isinstance(result, RatingResult) and result.question.is_grade_question and result.counts_for_grade
```
hmm, the `.count_sum` above suggests that this is even a PublishedRatingResult; I feel that we should add a type hint to the function, but this is already done in #2687

niklasmohrin commented on evap/results/tools.py:
```diff
@@ -397,7 +413,12 @@ def calculate_average_distribution(evaluation):
+                    isinstance(result, RatingResult) and result.counts_for_grade
+                    for result in questionnaire_result.question_results
+                )
+            if not questionnaire_result.questionnaire.is_dropout:
```
Why do we still need this branch? I thought that the exclusion of these results is now done using just the counts_for_grade attribute

jooooosef replied:
correct, I think this came from a rebase mistake on my side

niklasmohrin commented on evap/staff/tests/test_forms.py:
```diff
@@ -1150,3 +1152,59 @@ def test_save_makes_a_copy(self):
+
+
+class QuestionFormTests(TestCase):
+    def test_allows_additional_textanswers_persistence_after_type_change(self):
```
can you change the type back at the end and assert that the bool is updated as expected, even if we don't explicitly set it?

niklasmohrin commented on evap/staff/tests/test_live.py:
```diff
@@ -247,6 +247,114 @@ def test_questionnaire_selection(self):
+                additional_textanswers,
+                bool(
+                    row.find_element(By.CSS_SELECTOR, "input[id$='-allows_additional_textanswers']").get_attribute(
+                        "disabled"
```
Is this correct for elements with `disabled=""` (which I believe is what the browser inserts if you just write `disabled`)

Can we get a `is None` in there?

edit: ugh, just look at the source code of getAttribute, we just call into minified js that does all kinds of stupid stuff. I think we just want to use `getProperty` or `isEnabled`. The selenium docs are not very helpful with what "enabled" means in this context, but I think it means what we want here. Please investigate

jooooosef replied:
The browser does insert `disabled=""` here.
According to the [Selenium Docs](https://www.selenium.dev/selenium/docs/api/py/selenium_webdriver_remote/selenium.webdriver.remote.webelement.html#selenium.webdriver.remote.webelement.WebElement.get_attribute) this is `None` if the `disabled` is missing, so we could get a `None` here. 

But we can use [is_enabled](https://www.selenium.dev/selenium/docs/api/py/selenium_webdriver_remote/selenium.webdriver.remote.webelement.html#selenium.webdriver.remote.webelement.WebElement.is_enabled) to check the actual state.

niklasmohrin commented on evap/staff/tests/test_live.py:
```diff
@@ -247,6 +247,114 @@ def test_questionnaire_selection(self):
+
+        # Change to Dropout
+        self.set_tomselect_value(questionnaire_type_select, str(Questionnaire.Type.DROPOUT))
+        self.assertTrue(row.find_element(By.CSS_SELECTOR, "input[id$='-counts_for_grade']").get_attribute("disabled"))
```
shouldn't we also check that the value has changed, not only the "disabled" state?

niklasmohrin commented on evap/staff/tests/test_live.py:
Very cool tests still, looks complicated but actually easy to follow!

jooooosef replied:
Thanks :^)

niklasmohrin commented on evap/staff/tests/test_views.py:
```diff
@@ -3048,6 +3048,7 @@ def test_num_queries_is_constant(self):
             questionnaire=iter(questionnaires),
             question__type=QuestionType.TEXT,
             question__allows_additional_textanswers=False,
+            counts_for_grade=False,
```
hm, annoying. I wonder if we should set `default` on the counts_for_grade field to be a callable which picks the correct value based on the type. But maybe that makes things more complicated, don't know

niklasmohrin replied:
@richardebeling what do you think about callables for Field.default and Field.db_default? https://docs.djangoproject.com/en/6.0/ref/models/fields/#django.db.models.Field.default

richardebeling replied:
I'd be in favor of using that if it helped us, but I think it can't: The callable will not have the values of the other fields, right?

I don't think they will support that in the near future either, django would have to have a defined field value derivation order, we run into the same issue we have with the settings, there could be arbitrary complex data dependency chains here.

niklasmohrin commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,115 @@
+
+    constructor(questionTable: HTMLTableElement) {
+        this.questionTable = questionTable;
+        this.questionnaireTypeSelect = selectOrError<HTMLSelectElement>("#id_type");
```
How do we know this id? Can we use something more robust like selecting something from inside the table or so?

jooooosef replied:
The id is an automatically generated one from django.
We cannot select from inside the table, as this is outside the table, but needed for the logic (its at the top in the general options).
But I am now using an HTML attribute which we add to the form via the `widgets` in `forms.py`

niklasmohrin commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,115 @@
+            const checkboxes = questionTypeCell.querySelectorAll("input[type=checkbox]");
+
+            if (questionType === QUESTION_TYPE_TEXT || questionType === QUESTION_TYPE_HEADING) {
+                this.disableAndUncheckAll(checkboxes);
```
if you change the definition of `checkboxes` to use `querySelectorAll<HTMLInputElement>` then you can just write
```suggestion
                checkboxes.forEach(this.disableAndUncheck);
```
and delete the method completely

niklasmohrin commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,115 @@
+                    }
+                });
+            } else {
+                this.enableAndInitAll(checkboxes, true);
```
same here, I would just do
```suggestion
                checkboxes.forEach(box => this.enableAndInit(box, true));
```

niklasmohrin commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,115 @@
+    };
+
+    private initialize = () => {
+        // Initialize the state of all checkboxes based on current question types and questionnaire type
```
So do we still need to set the disabled attribute in the Django form? no, right? if we can, I would be very happy if we can delete this relic of the past

jooooosef replied:
YEEEEEEEEEEEEES

niklasmohrin commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,115 @@
+    private handleQuestionTypeChange = (e: Event) => {
+        const target = e.target as HTMLElement;
+        const questionTypeCell = target.closest("td.question-type");
+        if (questionTypeCell && target.matches("select")) {
```
can you change this to `if (!something) return`?

niklasmohrin reviewed (commented):
great stuff, some clean up work but I think we are close to done :)

jooooosef committed:
some typing and cleaner code

jooooosef committed:
linter unhappy

jooooosef committed:
frontend things made better

jooooosef committed:
fix tests

jooooosef committed:
unnessessary typecast

richardebeling commented on evap/evaluation/tests/test_models.py:
The changes in this file seem to be 100% about "allows additional text answers", not about "counts_for_grade". Is this intended, or just a rebase f-up?

jooooosef replied:
Yes, it is now only about the `allows additional text answers`, since we moved `counts_for_grade` into the `QuestionAssignment` (I think I originally wrote this). The first part (with `test_locked_contributor_questionnaire`) seems to be from the rebase. 
But the part with the `update_fields` from the second test is new, I cant find anything testing that. So I think I would like to keep this part, I have removed the rest.

richardebeling replied:
(note: I wasn't trying to say "This is bad", I was just trying to say "This doesn't seem like it belongs into this PR here", without even having looked at the code in detail. If there were added tests here that we want to get merged, opening a separate PR with these changes would be nice. The changes here are fine now, but in the interest of getting code merged quickly, it can be helpful to move such drive-by changes out into separate PRs)

jooooosef committed:
remove double testing

jooooosef committed:
fix eslint

niklasmohrin commented on evap/staff/tests/test_live.py:
```diff
@@ -247,6 +247,115 @@ def test_questionnaire_selection(self):
+    def test_question_type_disabling_logic(self):
+        def assert_type_allows(row, type_select, question_type, additional_textanswers, counts_for_grade):
+            self.set_tomselect_value(type_select, str(question_type))
+            self.assertEqual(
+                additional_textanswers,
+                row.find_element(By.CSS_SELECTOR, "input[id$='-allows_additional_textanswers']").is_enabled(),
+            )
+            self.assertEqual(
+                counts_for_grade,
+                row.find_element(By.CSS_SELECTOR, "input[id$='-counts_for_grade']").is_enabled(),
+            )
```
nit: we usually put the expected value second
```suggestion
            self.assertEqual(
                row.find_element(By.CSS_SELECTOR, "input[id$='-allows_additional_textanswers']").is_enabled(),
                additional_textanswers,
            )
            self.assertEqual(
                row.find_element(By.CSS_SELECTOR, "input[id$='-counts_for_grade']").is_enabled(),
                counts_for_grade,
            )
```

niklasmohrin commented on evap/staff/tests/test_live.py:
```diff
@@ -247,6 +247,115 @@ def test_questionnaire_selection(self):
+
+        new_row = self.selenium.find_elements(By.CSS_SELECTOR, "#question_table tbody tr")[
+            -2
+        ]  # the last row is the add row button
```
```suggestion
        new_row = self.selenium.find_elements(By.CSS_SELECTOR, "#question_table tbody tr")[1]
```

niklasmohrin commented on evap/staff/tests/test_live.py:
nice

niklasmohrin commented on evap/staff/templates/staff_questionnaire_form.html:
```diff
@@ -88,6 +95,10 @@ <h5 class="card-title">{% translate 'Questions' %}</h5>
+            import { StaffQuestionnaireForm } from "{% static 'js/staff-questionnaire-form.js' %}";
+            
+            const staffQuestionnaireForm = new StaffQuestionnaireForm(document.getElementById("question_table"));
+            
```
there is trailing whitespace on the empty lines here

niklasmohrin commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,104 @@
+        this.questionTable.addEventListener("change", this.handleQuestionTypeChange);
+        this.questionnaireTypeSelect.addEventListener("change", this.handleQuestionnaireTypeChange);
+
+        this.initialize();
```
I would either inline this method, or not call it here and instead let the caller run this. We have this semi pattern, for example in the contact modal, that `constructor` just queries the DOM and a second method, in this case `attach`, must be called to activate the component. This would also include the addEventListener` above.

I don't have a strong opinion, we can do everything in the constructor (without the extra method), or we move the event listeners to initialize and rename this accordingly and let the caller call this second function too.

jooooosef replied:
used attach

niklasmohrin commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,104 @@
+
+    constructor(questionTable: HTMLTableElement) {
+        this.questionTable = questionTable;
+        const form = selectOrError<HTMLFormElement>("#questionnaire-form");
```
I was wondering when seeing the caller why this was called StaffQuestionnaireForm but the only argument is the question table. So here we see that it does indeed use the entire questionnaire form. So let's make the caller pass this in

niklasmohrin commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,104 @@
+    private readonly questionnaireTypeSelect: HTMLSelectElement;
+
+    constructor(questionTable: HTMLTableElement) {
+        this.questionTable = questionTable;
```
You can merge the declaration and immediate assignment like this:
```typescript
cosntructor(
    private readonly questionTable: HTMLTableElement,
    // ...
)
```

niklasmohrin commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,104 @@
+            });
+        } else {
+            checkboxes.forEach(checkbox => this.enableAndInit(checkbox, true));
+        }
```
This contains some shortcuts that I think make it more difficult than it needs to be, I think a simpler version would be this:
```suggestion
        const questionType = saneParseInt(questionTypeSelect.value);
        const questionnaireType = saneParseInt(this.questionnaireTypeSelect.value);

        for (const checkbox of questionTypeCell.querySelectorAll<HTMLInputElement>("input[type=checkbox]")) {
            const shouldDisableFromQuestion = questionType === QUESTION_TYPE_TEXT || questionType === QUESTION_TYPE_HEADING;
            const shouldDisableFromQuestionnaire = checkbox.classList.contains("counts-for-grade-checkbox") && questionnaireType == QUESTIONNAIRE_TYPE_DROPOUT;

            if (shouldDisableFromQuestion || shouldDisableFromQuestionnaire) {
                this.disableAndUncheck(checkbox);
            } else {
                this.enableAndInit(checkbox, true);
            }
        };
```

niklasmohrin commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,104 @@
+        const selectedType = saneParseInt(this.questionnaireTypeSelect.value);
+        const countsForGradeCheckboxes = document.querySelectorAll(".counts-for-grade-checkbox");
+
+        countsForGradeCheckboxes.forEach(checkbox => {
```
and this establishes the same invariants as the loop in handleQuestionTypeChange, right? So maybe we should introduce a helper function `updateCheckboxes(checkboxes: HTMLCollection<HTMLInputElement>, questionnaire_type, question_type)` that does this loop?

niklasmohrin commented on evap/results/tools.py:
```diff
@@ -323,12 +337,17 @@ def avg_distribution(weighted_distributions):
-            (unipolarized_distribution(result), result.count_sum)
+            (
+                unipolarized_distribution(cast("PublishedRatingResult", result)),
+                cast("PublishedRatingResult", result).count_sum,
```
@richardebeling We are going to have to keep an eye on this one. Clearly we currently have that whenever a RatingResult comes by here, then it is actually a PublishedRatingResult. But that is just because we create RatingResult only if evaluation.can_publish_rating_results is False and we _also_ don't call this function because that implies that evaluation.can_publish_grade_results is False. This is all fragile, I don't like it, but this is not a concern for this PR

richardebeling replied:
would `assert isinstance`, maybe with a short comment explaining the reasoning, make more sense then?

jooooosef replied:
yes its not a comphrension now anymore as I have added the assert statement

niklasmohrin commented on evap/results/tools.py:
```diff
@@ -397,8 +418,12 @@ def calculate_average_distribution(evaluation):
+                assert not any(
+                    isinstance(result, RatingResult) and result.counts_for_grade
+                    for result in questionnaire_result.question_results
+                )
```
If this ever fails, we will be happy to have added a message like
```suggestion
                ), f"Dropout questionnaire {questionnaire_result.questionnaire.id} has results that count"
```

niklasmohrin commented on evap/results/tools.py:
```diff
@@ -397,8 +418,12 @@ def calculate_average_distribution(evaluation):
         for questionnaire_result in contribution_result.questionnaire_results:
-            if not questionnaire_result.questionnaire.is_dropout:  # dropout questionnaires are not counted
-                grouped_results[contribution_result.contributor].extend(questionnaire_result.question_results)
+            if questionnaire_result.questionnaire.is_dropout:  # dropout questionnaires are not counted
```
```suggestion
            if questionnaire_result.questionnaire.is_dropout:
```

niklasmohrin commented on evap/results/tools.py:
```diff
@@ -417,6 +442,8 @@ def calculate_average_distribution(evaluation):
                     ]
                 ),
+                # The weight of this contributors grade is supposed to represent the number of students the
+                # contributor interacted with, which we derive from the max answer count, independently of counts_for_grades.
```
```suggestion
                # contributor interacted with, which we derive from the max answer count, independently of counts_for_grade.
```

niklasmohrin commented on evap/results/tests/test_tools.py:
```diff
@@ -454,10 +456,105 @@ def test_dropout_questionnaires_are_not_included(self):
+        dropout_assignment.save()
         cache_results(self.evaluation)
 
-        calculated_grade = distribution_to_grade(calculate_average_distribution(self.evaluation))
-        self.assertAlmostEqual(calculated_grade, 1.5)
```
Can you add the removed lines here back to the existing test?

niklasmohrin commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,104 @@
+        if (questionnaireType === QUESTIONNAIRE_TYPE_DROPOUT) {
+            checkboxes.forEach(checkbox => {
+                if (checkbox.classList.contains("counts-for-grade-checkbox")) {
+                    this.disableAndUncheck(checkbox);
```
(out of scope for this PR)

@janno42 It would probably be a nice feature to add a tooltip to the disabled checkboxes that display the reason for why the checkbox is disabled ("Questions of a dropout questionnaire cannot count towards the evaluation's grade" or so)). I think this is slightly more complicated than "very small" or maybe even "small" because this typescript logic here has to interact with bootstrap tooltips and changing labels, and the knowledge of when things are disabled leaks out of this file, but it should be doable.

niklasmohrin reviewed (commented):
Very cool! I am pretty happy with the state of this and with what I was reading, good job. I have a handful of mostly nits, except for the typescript code which I think can use another touch.

I believe that our current plan is to make another release soon with the fixes for the recently discovered bugs, and I think it makes sense to hold this PR back until then, but after that I don't see anything in the way of merging.

jooooosef committed:
Coerce counts_for_grade in QuestionAssignment.save

enforce counts_for_grade on model level

jooooosef commented:
I just realised that in the `test_dropout_questionnaire_excluded_from_distribution` test we had:
 ```python
dropout_assignment.counts_for_grade = True
dropout_assignment.save()
```
This means, that we can actually save wrong assignemnts, which we should prevent in the first place. So I added a save overrite to the QuestionAssignment model.

Also: in the `QuestionnaireVotingForm` I added a `select_related(question)`, so we have it all in one SQL query at the beginning.

jooooosef committed:
avoid n+1 queries

direclty fetch question in the first call, so we dont have to requery the db afterward every time

jooooosef committed:
tidy up some stuff

jooooosef committed:
reject invalid grade-counting assignments
- when wrong values are submitted we now actively reject the request
- also add tests

jooooosef committed:
raise integrity error in QuestionAssignment model when counts_for_grade is set wrongly

also, and unfortunatly, fix tests

richardebeling commented on evap/evaluation/tests/test_models.py:
```diff
@@ -91,6 +92,40 @@ def test_questionnaire_cascading_delete_gc(self):
+            QuestionAssignment, questionnaire=dropout, question=grade_question, counts_for_grade=False
+        )
+        QuestionAssignment.objects.filter(pk=dropout_assignment.pk).update(counts_for_grade=True)
+        dropout_assignment.refresh_from_db()
```
I'm a bit torn on this one, because it creates a setup that we consider impossible, and tests our behavior in that situation -- it opens up the question "why do we care about our behavior in an impossible situation?". People reading the test may infer that this is considered valid state.

Do we gain anything if we do this, compared to just doing `assignment.counts_for_grade = True; assignment.save()`? I think it triggers the exact same code path, right? In that case, I'd just do that.

richardebeling commented on evap/staff/templates/staff_questionnaire_form.html:
```diff
@@ -66,6 +66,13 @@ <h5 class="card-title">{% translate 'Questions' %}</h5>
                                         </div>
+                                        <div class="form-check mt-1">
+                                            <input class="form-check-input counts-for-grade-checkbox" id="{{ form_element.counts_for_grade.id_for_label }}"
+                                                   name="{{ form_element.counts_for_grade.html_name }}" type="checkbox" data-keep="true"
```
what do we need data-keep="true" for?

richardebeling replied:
More general: Why do we need to custom-render this checkbox, why can't we just render the form field (using `bootstrap_form_field_widget.html`)?

jooooosef replied:
For `data-keep="true"`:
When we omit that, using the "add-another" button would leave the checkboxes unchecked. 
In the implementation of the `sortable-form.ts` is explicitly stated, that in order to avoid the error from #644 we need to empty the formset but since we want the checkboxes to be checked initially we have to add data-keep here. 

For `bootstrap_form_field_widget.html`:
When we use our `checkbox.html` we have no option to set the label, as this is empty:
```html
<div class="form-check">
    <input class="form-check-input" type="{{ widget.type }}" name="{{ widget.name }}"
        {% if widget.value != None %} value="{{ widget.value|stringformat:'s' }}"{% endif %}
        {% include 'django/forms/widgets/attrs.html' %} />
    <label class="form-check-label" for="{{ widget.attrs.id }}"></label> 
</div>
```
But we want a label here so we cant use that

richardebeling commented on evap/staff/tests/test_forms.py:
```diff
@@ -598,7 +600,11 @@ def test_prevent_contribution_deletion_with_answers(self):
+        baker.make(
+            RatingAnswerCounter,
+            contribution=contribution,
+            assignment__question__type=QuestionType.POSITIVE_LIKERT,
```
This PR changes a bunch of tests to explicitly specify `QuestionType.POSITIVE_LIKERT` instead of letting baker choose the value -- the intention here is always to ensure that the default values for `allows_additional_text_answers` and `counts_for_grades` (both True) do not make the test fail in our custom .save() method, right?

This also means that if we make baker deduce correct default values for these two bools in the future, we will have lost test coverage here, because these tests now only cover the features with POSITIVE_LIKERT questions, and not with other question types anymore. I think we probably want to revert this, as soon as we can (=> as soon as we have a baker version that can deduce the bools correctly)

jooooosef replied:
This is correct.
Should I make an issue for that?
Probably yes since you mention it in https://github.com/e-valuation/EvaP/pull/2450#discussion_r3610763050

richardebeling commented on evap/staff/tests/test_forms.py:
```diff
@@ -1150,3 +1156,71 @@ def test_save_makes_a_copy(self):
+        Test that allows_additional_textanswers can be changed and persisted correctly
+        even after changing the question type from TEXT/HEADING to a Likert type.
+        Regression test for #2539.
+        """
```
This is unrelated to non-counting questions. Please, for our sanity, try to keep these kinds of changes in separate PRs. It's significantly more confusing when looking at Git history, finding the GitHub PRs is harder, and it just prolongs review processes.

jooooosef replied:
Yes ofc.
I have added this test before #2539 was merge bc this would have also been the case for "counts_for_grade" so we decided to fix this bug and add counts_for_grade 

EDIT: Its actually covered by the test `evap/staff/tests/test_views.py:3408` so I will remove it

richardebeling commented on evap/static/ts/src/staff-questionnaire-form.ts:
```diff
@@ -0,0 +1,95 @@
+            } else {
+                this.enableAndInit(checkbox, true);
+            }
+        };
```
This manipulates all checkboxes, even unrelated ones we may add in the future. Can/should we add some filter that says "these are the checkboxes we want to modify"? I think that would also help code readability. Currently, I can infer that one of the checkboxes we care about about counts-for-grade, but this code does not say anything about how many other checkbox kinds there are or what they are for.

jooooosef replied:
I think I made it more readable and easier to understand now :)

richardebeling commented on evap/staff/tests/test_live.py:
```diff
@@ -249,6 +252,113 @@ def test_questionnaire_selection(self):
+        self.set_tomselect_value(questionnaire_type_select, str(Questionnaire.Type.DROPOUT))
+        self.assertFalse(row.find_element(By.CSS_SELECTOR, "input[id$='-counts_for_grade']").is_enabled())
+        self.assertFalse(row.find_element(By.CSS_SELECTOR, "input[id$='-counts_for_grade']").is_selected())
+        self.assertTrue(row.find_element(By.CSS_SELECTOR, "input[id$='-allows_additional_textanswers']").is_enabled())
```
Let's also assert the "enabled" state of this here and below.

richardebeling commented on evap/staff/tests/test_live.py:
```diff
@@ -249,6 +252,113 @@ def test_questionnaire_selection(self):
+
+        self.wait.until(
+            lambda driver: len(driver.find_elements(By.CSS_SELECTOR, "#question_table tbody tr.sortable")) >= 2
+        )
```
This should be a no-op -- what do we need it for?

jooooosef replied:
I tried to be sure, that our JS Code ran and still did not change the invisible extra row, but that didnt quite work.
I now changed the the condition to wait until the initially enabled checkbox is  disabled, to be sure the JS ran and then do the assertions

richardebeling commented on evap/staff/tests/test_live.py:
```diff
@@ -249,6 +252,113 @@ def test_questionnaire_selection(self):
+        counts_for_grade_checkbox = extra_row.find_element(By.CSS_SELECTOR, "input[id$='-counts_for_grade']")
+
+        self.assertEqual(type_select.get_attribute("value"), "")
+        self.assertTrue(counts_for_grade_checkbox.is_enabled())
+        self.assertTrue(counts_for_grade_checkbox.is_selected())
```
what about allows_additional_textanswers?

richardebeling commented on evap/staff/tests/test_views.py:
```diff
@@ -782,7 +782,11 @@ def test_textanswer_review_state_tags(self, textanswer_review_state_mock):
+        baker.make(
+            TextAnswer,
+            contribution=evaluation.general_contribution,
+            assignment__counts_for_grade=False,
```
Please create a follow-up ticket about making `baker.make` deduce the correct default value. People trying to write tests should get a sane default here.

I think we can just make all our tests do `from evap.testing.tools import baker`, with our baker instance just injecting the correct default value.

jooooosef replied:
Yes see #2761

jooooosef committed:
remove duplicated test

- see test_can_change_question in staff/test_views.py

jooooosef committed:
refactor ts for the questionnaire form table

jooooosef committed:
add an assert and some smaller readability / spelling

jooooosef committed:
make tests better

jooooosef committed:
format the html since we finally have a formatter (yay)

niklasmohrin commented on evap/staff/tests/test_views.py:
```diff
@@ -3381,8 +3419,37 @@ def test_can_change_question(self) -> None:
+            counts_for_grade=True,
+        )
+
+        page = self.app.get(f"/staff/questionnaire/{questionnaire.pk}/edit", user=self.manager)
```
should use `reverse`

niklasmohrin commented on evap/evaluation/models.py:
```diff
@@ -1303,7 +1303,7 @@ class Meta:
                 ),
                 name="check_evaluation_textanswer_or_heading_question_has_no_additional_textanswers",
-            )
+            ),
```
pls revert
